### PR TITLE
blog on dynamic adversarial data collection space with mnist example

### PR DESCRIPTION
**Pull request by chrisemezue**

### CODE REVIEW

**Comment by chrisemezue** [comment]
Am also working concurrently on this [HackMD](https://hackmd.io/@wzo-xq9VRn-8ZBXawr4mvA/ryRBXW595)

File: mnist-adversarial.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+Nevertheless, it has been shown that these SOTA models still find it difficult to correctly predict human handwriting of the digits: largely due to the distrbution shift in the way humans actually write and the data in the test set.
+
+Therefore these models need humans in the loop to provide them with adversarial data in order to make them more robust and generalize better.
+
+In this short guide, I will walk you through creating an adversarial space for your MNIST model. This space will enable users to try to fool your model by providing examples that are hard for the model. Finally, the model will be trained on these adversarial samples after a certain number of them have been collected. The cycle continues for as many rounds as possible. Researchers<sup>[1](https://aclanthology.org/2022.findings-acl.18.pdf) </sup>have shown that to get the best out of DADC, one should run it for many rounds (as against only 1-3 rounds).
```
Comment (osanseviero) [inline mnist-adversarial.md:27]:
```suggestion
In this short guide, I will walk you through creating an adversarial Space for your MNIST model. This Space will enable users to try to fool your model by providing examples that are hard for the model. Finally, the model will be trained on these adversarial samples after a certain number of them have been collected. The cycle continues for as many rounds as possible. Researchers<sup>[1](https://aclanthology.org/2022.findings-acl.18.pdf) </sup>have shown that to get the best out of DADC, one should run it for many rounds (as against only 1-3 rounds).
```

File: mnist-adversarial.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+Below is a simple space to interact with the `MNIST_Model` which I trained up till 98% accuracy on the test set. You draw a number and see the model's prediction. The space is [here](https://huggingface.co/spaces/chrisjay/simple-mnist-classification). Try to fool this model😁. Use your funniest handwriting; write on the sides of the canvas; go wild!
+
+<iframe src="https://hf.space/embed/chrisjay/simple-mnist-classification/+" frameBorder="0" width="100%" height="360px" title="Gradio app" allow="accelerometer; ambient-light-sensor; autoplay; battery; camera; document-domain; encrypted-media; fullscreen; geolocation; gyroscope; layout-animations; legacy-image-formats; magnetometer; microphone; midi; oversized-images; payment; picture-in-picture; publickey-credentials-get; sync-xhr; usb; vr ; wake-lock; xr-spatial-tracking" sandbox="allow-forms allow-modals allow-popups allow-popups-to-escape-sandbox allow-same-origin allow-scripts allow-downloads"></iframe>
+Currently the model is not trained. Therefore we expect inaccurate predictions. We will proceed to train this model with adversarially collected data.
```
Comment (osanseviero) [inline mnist-adversarial.md:68]: This line is a bit confusing since above you mentioned the model was trained up to 98% accuracy

File: mnist-adversarial.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+### Flagging the model
+  
+Were you able to fool the model above?😀 Now it's time to _flag_ your adversarial example. __Flagging__ refers to the processes that happen when a user is able to fool the model. Generally flagging involves saving the adversarial example, and other custom things you want to implement. For our flagging, we will define a function to: 
+1. save the adversarial example to a dataset
+2. train the model on the adversarial example after some threshold samples have been collected.
+3. repeat 1-2 many times. 
+
+>Note: Gradio has a built-in flaggiing system that allows you easily flag adversarial samples of your model. Read more about it [here](https://gradio.app/using_flagging/).
```
Comment (osanseviero) [inline mnist-adversarial.md:77]:
```suggestion
>Note: Gradio has a built-in flagging system that allows you easily flag adversarial samples of your model. Read more about it [here](https://gradio.app/using_flagging/).
```

File: mnist-adversarial.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+Here is an explanation of what my `flag` function does. For more details feel free to peruse the full code [here](https://huggingface.co/spaces/chrisjay/mnist-adversarial/blob/main/app.py#L314). 
+
+```python=
+def flag(input_image,correct_result,adversarial_number):
+     """
+    It takes in an image, the correct result, and the number
+    of adversarial images that have been uploaded so far.
+    It saves the image and metadata to a local directory, uploads
+    the image and metadata to the hub, and then pulls the data from
```
Comment (osanseviero) [inline mnist-adversarial.md:87]:
```suggestion
    the image and metadata to the Hub as a dataset repository, and then pulls the data from
```

File: mnist-adversarial.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    the hub to the local directory.
```
Comment (osanseviero) [inline mnist-adversarial.md:88]:
```suggestion
    the Hub to the local directory. 
```

File: mnist-adversarial.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    that have been uploaded to the dataset
+    
+    :return: The output is the output of the flag function.
+    """
+
+```
+
+### Putting it all together
+
+The final step is to put all the components together into one space! I have created the [MNIST Adversarial](https://huggingface.co/spaces/chrisjay/mnist-adversarial) space here for dynamic adversarial data collection for the MNIST handwritten recognition task.
```
Comment (osanseviero) [inline mnist-adversarial.md:105]:
```suggestion
The final step is to put all the components together into one Space! I have created the [MNIST Adversarial](https://huggingface.co/spaces/chrisjay/mnist-adversarial) space here for dynamic adversarial data collection for the MNIST handwritten recognition task. 
```

**Review by osanseviero — COMMENTED** [review]
Very very cool blog post! Looking forward to the final version :fire: I left mostly nits

File: mnist-adversarial.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
+# How to train a model using community adversarial data
+> Still working on the title...feel free to suggest alternatives.
+
+##### What you will learn here
+- 💡the idea of dynamic adversarial data collection and why it is important.
+- ⚒how to collect adversarial data dynamically and train your model on them - using an MNIST handwritten digit recognition task as an example. 
+
+
+## Dynamic adversarial data collection (DADC)
```
[the remark this answers is not in the document]
Comment (chrisemezue) [reply mnist-adversarial.md:9]: thanks

File: mnist-adversarial.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
+# How to train a model using community adversarial data
```
[the remark this answers is not in the document]
Comment (chrisemezue) [reply mnist-adversarial.md:1]: thanks. will do

[the remark this answers is not in the document]
Comment (chrisemezue) [reply mnist-adversarial.md:1]: @osanseviero is there a size specification for the image in ?
```
https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1593126474392-5ef50182b71947201082a4e5.jpeg?w=200&h=200&f=face
```

[the remark this answers is not in the document]
Comment (osanseviero) [reply mnist-adversarial.md]: It's usually the image from https://huggingface.co/chrisjay. In your case it's https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1634562259269-5fbfa21ec153e968e63deb56.jpeg?w=200&h=200&f=face

[the remark this answers is not in the document]
Comment (chrisemezue) [reply mnist-adversarial.md]: thanks

File: mnist-adversarial.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+In this short guide, I will walk you through creating an adversarial space for your MNIST model. This space will enable users to try to fool your model by providing examples that are hard for the model. Finally, the model will be trained on these adversarial samples after a certain number of them have been collected. The cycle continues for as many rounds as possible. Researchers<sup>[2](https://aclanthology.org/2022.findings-acl.18.pdf) </sup>have shown that to get the best out of DADC, one should run it for many rounds (as against only 1-3 rounds).
+
+This part will be divided into the following sections:
+1. Configuring your model
+2. Interacting with your model
+3. Flagging your model
+4. Putting it all together
+
+### Configuring your model
+First of all, you need to define your model architecture. My simple model architecture below is made up of two convolutional networks connected to a 50 dimensional fully connected layer and the final layer for the 10 classes.
```
Comment (douwekiela) [inline mnist-adversarial.md:65]: and _a_ final layer?

File: mnist-adversarial.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+<iframe src="https://hf.space/embed/chrisjay/simple-mnist-classification/+" frameBorder="0" width="100%" height="360px" title="Gradio app" allow="accelerometer; ambient-light-sensor; autoplay; battery; camera; document-domain; encrypted-media; fullscreen; geolocation; gyroscope; layout-animations; legacy-image-formats; magnetometer; microphone; midi; oversized-images; payment; picture-in-picture; publickey-credentials-get; sync-xhr; usb; vr ; wake-lock; xr-spatial-tracking" sandbox="allow-forms allow-modals allow-popups allow-popups-to-escape-sandbox allow-same-origin allow-scripts allow-downloads"></iframe>
+
+### Flagging your model
+  
+Were you able to fool the model above?😀 Now it's time to _flag_ your adversarial example. __Flagging__ refers to the processes that happen when a user is able to fool the model. For our flagging, we will define a function to: 
+1. save the adversarial example to a dataset
+2. train the model on the adversarial example after some threshold samples have been collected.
```
Comment (douwekiela) [inline mnist-adversarial.md:103]: on the adversarial example_s_ I think

File: mnist-adversarial.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+3. repeat 1-2 many times. 
+
+>Note: Gradio has a built-in flaggiing callback that allows you easily flag adversarial samples of your model. Read more about it [here](https://gradio.app/using_flagging/).
+
+Here is an explanation of what my `flag` function does. For more details feel free to peruse the full code [here](https://huggingface.co/spaces/chrisjay/mnist-adversarial/blob/main/app.py#L314). 
+
+```python=
```
Comment (douwekiela) [inline mnist-adversarial.md:110]: To me this feels a bit weird in a code block, while it's basically just comments? Can you not just write it out in plain text?

File: mnist-adversarial.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
+---
+title: "How to train a model using community adversarial data"
+thumbnail: /blog/assets/20_accelerate_library/accelerate_diff.png
+---
+
+<h1>
+    How to train a model using community adversarial data
```
Comment (douwekiela) [inline mnist-adversarial.md:7]: Tbh the title confuses me a little - it's not clear to me what "community adversarial data" means

**Review by douwekiela — COMMENTED** [review]
Very cool! Small comments

File: mnist-adversarial.md (hunk 2 above)
Comment (chrisemezue) [reply mnist-adversarial.md]: removed!

File: mnist-adversarial.md (hunk 9 above)
Comment (chrisemezue) [reply mnist-adversarial.md:65]: thanks

File: mnist-adversarial.md (hunk 10 above)
Comment (chrisemezue) [reply mnist-adversarial.md:103]: where please @douwekiela ?

File: mnist-adversarial.md (hunk 12 above)
Comment (chrisemezue) [reply mnist-adversarial.md:7]: thanks @douwekiela 
is it enough to just change the title? or are you proposing explaining more on what "community adversarial" entails?

File: mnist-adversarial.md [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        return F.log_softmax(x)
+```
+
+Now that you have defined your model, you can go ahead with training your model on the standard MNIST train/dev dataset or use a pretrained model. 
+> Should I add some more code on training? the dataloader, train function, etc.
+
+
+### Interacting with your model
+
+Now that you have your model, you need a way for users to interact with it: specifically you want them to be able to draw a number and see the model's prediction. You can do all that with [🤗 Spaces](https://huggingface.co/spaces).
```
[the remark this answers is not in the document]
Comment (chrisemezue) [reply mnist-adversarial.md:93]: thanks. done

File: mnist-adversarial.md (hunk 11 above)
Comment (chrisemezue) [reply mnist-adversarial.md:110]: I want to show them the concept of the `flag` function. So they actually understand that it is a function....

Comment (chrisemezue) [reply mnist-adversarial.md:110]: is writing in plain still better?

Comment (chrisemezue) [reply mnist-adversarial.md:110]: I actually wanted to display the [whole function](https://huggingface.co/spaces/chrisjay/mnist-adversarial/blob/main/app.py#L333-L395) but it is too long.

File: mnist-adversarial.md (hunk 10 above)
Comment (chrisemezue) [reply mnist-adversarial.md:103]: thanks. done

File: mnist-adversarial.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrisemezue) [reply mnist-adversarial.md]: done

File: mnist-adversarial.md (hunk 12 above)
Comment (douwekiela) [reply mnist-adversarial.md:7]: Yeah I'd change it, maybe "How to train a model dynamically using adversarial data" or something along those lines?

**Review by osanseviero — APPROVED** [review]
The blog is great, I really like it! Once the space is ready I think we can merge this, nice work!

File: mnist-adversarial.md [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+<iframe src="https://hf.space/embed/chrisjay/mnist-adversarial/+" frameBorder="0" width="1000px" height="660px" title="Gradio app" allow="accelerometer; ambient-light-sensor; autoplay; battery; camera; document-domain; encrypted-media; fullscreen; geolocation; gyroscope; layout-animations; legacy-image-formats; magnetometer; microphone; midi; oversized-images; payment; picture-in-picture; publickey-credentials-get; sync-xhr; usb; vr ; wake-lock; xr-spatial-tracking" sandbox="allow-forms allow-modals allow-popups allow-popups-to-escape-sandbox allow-same-origin allow-scripts allow-downloads"></iframe>
+
+
+## Conclusion
+
+
+Dynamic Adversarial Data Collection (DADC) has been gaining traction in the NLP research community as a way to gather more robust datasets, and improve model evaluation and task performance. By allowing the dynamic collection of human-written data with models in the loop, we can improve the generalization potential of our models.
```
[the remark this answers is not in the document]
Comment (chrisemezue) [reply mnist-adversarial.md:114]: good point. Thanks for this.

File: .gitignore [hunk 15]
```diff
@@ -12,3 +12,6 @@ venv.bak/
 
 # Misc
 **/.DS_Store
+
+# VS Code 
+.vs/
```
Comment (osanseviero) [inline .gitignore:17]: I think you can remove this from `.gitignore`. You can have this in your local global `.gitignore` file though! https://stackoverflow.com/questions/45333484/how-to-use-global-gitignore-with-visual-studio-code

File: _blog.yml [hunk 16]
```diff
@@ -950,4 +950,15 @@
...
+
+
+- local: mnist-adversarial
+  title: "How to train your model dynamically using adversarial data"
+  author: chrisjay
+  thumbnail: /blog/assets/84_mnist_adversarial.JPG
+  date:  July 8, 2022
+  tags:
+  - mnist
+  - adversarial
```
Comment (osanseviero) [inline _blog.yml:964]: Maybe also add `guide` here

**Review by osanseviero — APPROVED** [review]
This is very nice blog post! I think you can solve the last couple of comments, resolve merge conflicts and then merge the PR to deploy it! 🔥 🚀

File: .gitignore (hunk 15 above)
Comment (chrisemezue) [reply .gitignore:17]: thanks

File: mnist-adversarial.md [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+title: "How to train your model dynamically using adversarial data"
+thumbnail: /blog/assets/84_mnist_adversarial.JPG
+---
+
+<h1>
+    How to train your model dynamically using adversarial data
+</h1>
+
+<div class="blog-metadata">
+    <small>Published July 8, 2022.</small>
```
[the remark this answers is not in the document]
Comment (chrisemezue) [reply mnist-adversarial.md]: done

**Comment by chrisemezue** [comment]
I have booked to post the blog tomorrow Saturday - 16th July. Is that okay @osanseviero ?

**Comment by osanseviero** [comment]
Sure, it's ok @chrisemezue! I will merge now so this gets deployed, but nobody gets automatic notifications out of it, so feel free to announce whenever you want!
